### PR TITLE
Enable access to remote mode solver from python

### PR DESCRIPTION
**Pull request by lucas-flexcompute**
The local mode solver can be used through `ModeSolver.solve_local`.

This is a WIP because the changes to the current mode solver can be "messy", so everyone's input is welcome.

The cache system had to be extended to allow for caching aliases, because the solve function returns the value of the `data_raw` property, which is cached. Thus `solve_local` must share a cache entry with `data_raw`, which is now possible by passing a name to the cache decorator.

### CODE REVIEW

**Review by tylerflex — COMMENTED** [review]
Looks good. I'm just trying to understand the ramifications of the new cacher / why that was needed.

1. all `cached_property` methods are unaffected?
2. what does `@cache(name="data_raw")` do? My understanding is that it caches the output of `obj.solve_local()` so it can be accessed by `obj.data_raw`? If my understanding is right, what's wrong with either the approach of:
```py
@cached_property
def data_raw(self):
    return self.solve_local()
```

or 
```py
def solve_local(self):
    return self.data_raw
```
?

**Comment by lucas-flexcompute** [comment]
> 1. all `cached_property` methods are unaffected?

Yes, it preserves the previous behavior. There's some decorator magic happening behind the scenes that allows us to use the decorator directly or as a function that creates a decorator (passing the `name` argument). So both of these work:
```python
@cache
def some_function(...):
    ...

@cache(name="cached name")
def other_function(...):
    ...
```

> 2. what does `@cache(name="data_raw")` do? My understanding is that it caches the output of `obj.solve_local()` so it can be accessed by `obj.data_raw`? If my understanding is right, what's wrong with either the approach of:
> 
> ```python
> @cached_property
> def data_raw(self):
>     return self.solve_local()
> ```
> 
> or
> 
> ```python
> def solve_local(self):
>     return self.data_raw
> ```

`data_raw` is the value that is cached as `"data_raw"` but it can be calculated with the local or remote solvers. Your first option eliminates the possibility of choosing, so it's not viable. The second doesn't work if `data_raw` has not yet been calculated, because once you get to the property, it will call the remote solver, but the user wanted to use the local one. The issue is that we cannot pass parameters to the property calculation to specify which solver to use. The solution here is to default to remote solver (so that's the one the property uses if we call it directly and it is not in the cache) and use the local solver only when the user calls `solve_local`. But we want that result to be cached as well, such that later they can use `data_raw` or `data` and get the local result, so what I did is to set `solve_local` to be cached as `data_raw` (like an alias). It also works the other way around: if `data_raw` has already been calculated, `solve_local` would return it directly (same for `solve`) regardless of how it was calculated when it was cached.

**Comment by tylerflex** [comment]
So is the idea that we want to provide a single property (stored as `.data_raw`) that can be computed with either `solve()` or `solve_local()`? and then both of those methods will either compute it with remote / local solver (respectively) or grab from the previously cached value?

I guess something I'm wondering is why not just introduce a `.data_raw_local` property that does the local solve and reserve `.data_raw` for remote solver jobs? It seems a bit confusing or unnecessarily complicated to store the remote / local in the same place. I probably just don't understand something though, but I was imagining the two calls `.solve()` and `solve_local()` call the remote and local mode solvers and have their own independent storage for the results, just to keep things separate and avoid complications. For example, note that if the `solve_local()` stores `.data_raw` and then a user calls `.solve()` they will get a different datastructure. So in some case this introduces a notion of state to the mode solver (different results depending on how the mode solver was used previously) and we would want to avoid this.

**Comment by lucas-flexcompute** [comment]
TBH I haven't been able to use the remote mode solver from python yet, so I don't know what data structure we get back (I assumed it would be the same as the local mode solver provides, but maybe I'm wrong). The idea of using the same attributes comes more from a simplicity in the UX, considering we want the user to use the remote solver, not the local one, due to the better accuracy. In that case, I imagine the user creating a (complex) script where they need the mode solver. During debug and exploratory phases, they can simply call `ModeSolver.solve_local()` immediately after creating the mode solver and all calls to `.data`, `.n_eff`, `.n_group`, and `.mode_area` would use that result. Once they are more or less happy with the script and want an accurate results, simply replace the line with `.solve()` or comment it out, since the default is the remote one, and nothing else needs to be changed.

But I understand what you're saying about state. In line with giving priority to the remote solver, we could have the call to `solve()` override the call to `local_solve()`, that is, have a separate structure for the local data and, if available, use that for the properties, otherwise, go through `data_raw`. We can redefine `ModeSolver.data_raw` to be something like:
```python
@property
def data(self):
    data = self._cached_properties.get("data_raw_local") or self.data_raw
    # do the symmetry expansion…
    return data
```
The only differences are that `data` is no longer cached and we're directly accessing the cache implementation for `data_raw_local`, which is something we have to remember if we ever change anything in the cache system (I don't think this is a big issue… we can write tests for this). Does it look better to you this way?

**Review by tylerflex — COMMENTED** [review]
This generally looks good to me. I also appreciate the clean up on the existing webapi functions. 

So it seems like `ModeSolver` will be local and `RemoteModeSolver` will be the the server-side one? I know at some point we discussed calling these `LocalModeSolver` and `ModeSolver`, respectively. Can you remind me why we do it this way? my worry is that we eventually would prefer users to use the remote one, so nudging them to this option seems preferable, but I guess there are some backwards compatibility issues?

File: tests/test_plugins/test_mode_solver.py [hunk 1]
```diff
@@ -21,7 +18,21 @@
...
+    mode_spec = td.ModeSpec(num_modes=3, target_neff=2.0)
+    _ = compute_modes(
+        eps_cross=[eps_cross] * 3,
+        coords=[coords, coords],
+        freq=td.C_0 / 1.0,
+        mode_spec=mode_spec,
+    )
+
+
+@pytest.mark.parametrize("MSClass", [ModeSolver])
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tests/test_plugins/test_mode_solver.py:34]: Yes, I was using the tests to test the actual remote solver, not only a mock-up. I'll add the mock response to have both classes tested by default, thanks!

**Comment by lucas-flexcompute** [comment]
> So it seems like `ModeSolver` will be local and `RemoteModeSolver` will be the the server-side one? I know at some point we discussed calling these `LocalModeSolver` and `ModeSolver`, respectively. Can you remind me why we do it this way? my worry is that we eventually would prefer users to use the remote one, so nudging them to this option seems preferable, but I guess there are some backwards compatibility issues?

Yes, after the discussion on Wednesday, I understood that we don't want to change the default behavior just yet, so keeping `ModeSolver` as the local version preserves compatibility with existing code. But since we're only merging this in 2.3, there's still time to rename both.

TBH I'm still not 100% conviced we should have 2 classes. They have no difference between them and, with the new `web.run` interface, it seems to me that there would be no confusion from the user side as to which solver they're using. What I mean is that we could just let `web.run` accept a `ModeSolver` instance and remove `RemoteModeSolver`.

If not, at least I'm going to add `ModeSolver.to_remote()` and `RemoteModeSolver.to_local()` to facilitate the interchange. I'm also adding the solver choice to the waveguide plugin.

**Comment by tylerflex** [comment]
> TBH I'm still not 100% conviced we should have 2 classes. They have no difference between them and, with the new web.run interface, it seems to me that there would be no confusion from the user side as to which solver they're using. What I mean is that we could just let web.run accept a ModeSolver instance and remove RemoteModeSolver.

Hm, in that case, would the `web.run(mode_solver)` just return a remotely calculated `ModeSolverData` whereas a `mode_solver.solve()` returns a locally computed `ModeSolverData`? If that seems simpler from an implementation standpoint, I think it would be fine to merge the two classes to handle it this way. We could consider also warning in `ModeSolver.solve()` that a new remote version exists now and that this will only compute local results (without sub pixel). If you think this makes sense then I'm 100% on board with it. It would be nice to have a cleaner namespace.

**Comment by lucas-flexcompute** [comment]
Sounds good! I think that makes the most sense!

**Review by tylerflex — APPROVED** [review]

**Pull request merged**